### ISSUE TITLE
Overlapped IO is not executed asynchronously

### ISSUE THREAD

**Issue by blschatz** [billziss-gh/winfsp#291]
## Bug Report

When using async reads on a file backed by a WinFSP UFS, the reads being issued to the UFS are being handled sequentially, rather than in parallel. 

### How to Reproduce

Here's a test program run against memfs
memfs-x64.exe -d "-1" -D - -F FOO -m \\.\v: -t 0 -s 1073741825

Setting a breakpoint in MemFS::Read() should yield multiple threads breaking, but only a single breakpoint is hit at a time. 

```
#include <iostream>
#include <Windows.h>
#include <fileapi.h>


int main()
{
    static long readOffset = 1024*1024 * 4;
    static int requests = 16;

	// generate a test file the first run
	if (GetFileAttributesW(L"v:\\foo.raw") == INVALID_FILE_ATTRIBUTES) {
		HANDLE whandle = ::CreateFileW(L"v:\\foo.raw", GENERIC_WRITE,
			0, NULL,
			CREATE_NEW, FILE_ATTRIBUTE_NORMAL, NULL);
		if (INVALID_HANDLE_VALUE == whandle)
		{
			fprintf(stderr,
				"error %d\n",
				GetLastError());
			exit(1);
		} // if ends

		void* src = malloc(16 * 1024 * 1024);
		for (long i = 0; i < (1024 / 16) + 1; i++) {
			::WriteFile(whandle, src, 16 * 1024 * 1024, NULL, NULL);
		}

		::CloseHandle(whandle);
	}

	// open the file and completion port
    HANDLE handle = ::CreateFileW(L"v:\\foo.raw", GENERIC_READ ,
        0, 0,
        OPEN_ALWAYS, FILE_ATTRIBUTE_NORMAL | FILE_FLAG_OVERLAPPED , 0);
    if (INVALID_HANDLE_VALUE == handle)
    {
		exit(1);
    } // if ends

	ULONG key = 1;
	HANDLE completionPort = CreateIoCompletionPort(handle, NULL, (ULONG_PTR)&key, 2);
	if (completionPort == NULL) {
		exit(1);
	}

	// issue 16 overlapped IO's
    void* buffers[16];
	OVERLAPPED ovr[16];

	int pendingCompletions = 0;
    for (int i = 0; i < 16; i++) {
		void* buffer = malloc(4096);
        buffers[i] = buffer;

        ovr[i].Offset = readOffset * i;
		ovr[i].OffsetHigh = 0;
		ovr[i].hEvent = NULL;

		boolean res = ReadFile(handle, buffer, 4096, NULL, &ovr[i]);
		int lastError = GetLastError();
		if (res) {
			// immediate read
			// no need to wait on overlapped

		}
		else {
			if (lastError == ERROR_IO_PENDING)
				pendingCompletions += 1;
		}
    }

	// collect the completed IO's
	LPOVERLAPPED CompletedOverlapped;

	for (int i = 0; i < pendingCompletions;  i++) {
        if (buffers[i] != NULL) {
			DWORD_PTR key;
			DWORD bytes;
			boolean bOK = GetQueuedCompletionStatus(completionPort, &bytes, &key, &CompletedOverlapped, INFINITE);

			if (!bOK) {
				fprintf(stderr,
					" %d\n",
					GetLastError());
				exit(1);
			}

		}
    }
}
```
### Behaviors

Expected is multiple concurrent reads in the UFS. Actual is a single read at a time being handled. 

### Environment

- OS version and build: Windows 10.1903
- WinFsp version : 2020 (1.6.20027)

**Comment by billziss-gh** [comment]
I have finally had the chance to look into this.

Some background:

The kernel mode FSD maintains an internal structure for every open file called a `FileNode`. This `FileNode` contains a shared-exclusive lock (i.e. a read-write lock), which it acquires in a shared manner for "read" operations (including READ, QUERY_INFORMATION, etc.) and exclusively for "write" operations (including WRITE, SET_INFORMATION, etc.)

The problem is that although a "read" operation is conceptually an operation that can happen concurrently with other "read" operations, nevertheless sometimes it has to update some internal state and for this to happen safely the lock needs to be momentarily acquired exclusively.

Often this pattern looks as follows (C pseudocode):

```C
NTSTATUS Read()
{
    AcquireExclusive(Resource);
    // update internal state
    ConvertExclusiveToShared(Resource);
    // send request to user mode file system
    return STATUS_PENDING;
}

VOID ReadComplete()
{
    // receive response from user mode file system 
    Release(Resource);
    // complete READ IRP
}
```

Now let us suppose that we have 4 threads A, B, C, D competing for a READ and that thread A acquires the lock first. Threads B, C, D end up waiting on the lock, while thread A holds the lock exclusively. After thread A updates any internal state it then executes `ConvertExclusiveToShared` to convert its lock from exclusive to shared. This appears to happen in a manner where threads B, C, D do not get a chance to wake up and compete for the lock. Thus thread A quickly gets the lock in shared state and threads B, C, D cannot proceed (because they also need the lock briefly in exclusive mode).

A potential fix is to change how the FSD converts locks from exclusive to shared. The FSD uses the Windows kernel [`ERESOURCE`](https://docs.microsoft.com/en-us/windows-hardware/drivers/kernel/introduction-to-eresource-routines) to implement its read-write locks and the [`ExConvertExclusiveToSharedLite`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-exconvertexclusivetosharedlite) DDI to convert them from exclusive to shared. This DDI appears to give priority to the thread that already owns the lock exclusively and we may want to consider a technique where the lock is instead released and reacquired shared (perhaps even doing acquisition using the [`ExAcquireSharedWaitForExclusive`](https://docs.microsoft.com/en-us/previous-versions/ff544370(v=vs.85)) DDI in order to give clear preference to exclusive waiters).

This is of course a potentially dangerous change and we have to think it through carefully. The good news is that there is a single place where the code needs to be changed: [`FspFileNodeConvertExclusiveToSharedF`](https://github.com/billziss-gh/winfsp/blob/v1.8B1/src/sys/file.c#L498-L512)

Thank you for identifying this problem and apologies that it took so long to look into. My not-so-good excuse is that I had verified that concurrent READ's were possible early on during WinFsp's development. However any number of things may have changed since that time within WinFsp (or even Windows) or I may have simply misinterpreted those original results.

**Comment by billziss-gh** [comment]
Commit 0296502 should resolve this issue.

Please note that your test needs a slight modification in order to show that this is the case. The problem is that when both buffered and non-buffered accesses are sent to a file, the FSD is forced to flush the file to avoid cache consistency problems. This means that it still needs to acquire exclusive locks during READ, which looks from the perspective of the user mode file system that READ's are serialized (rather than concurrent as they should be).

To avoid this problem during your testing you must do the following:

- Modify your test program to open files with the flag `FILE_FLAG_NO_BUFFERING`. Both uses of `CreateFile` must be modified. I include below a modified copy of your program for your convenience. (Please note: if you omit the `FILE_FLAG_NO_BUFFERING` flag, READ's will be satisfied from system cache and you will not be able to confirm that concurrent READs are possible.)

- Disable your Antivirus and any other filters. For example, if you are running Windows Defender it will access the `foo.raw` file in a way that will cause the FSD to flush the file.

- Start the file system with: `memfs-x64.exe -d -1 -D - -F FOO -m \\.\v: -M 10000`. The `-M 10000` option delays I/O completion up to 10 seconds within MEMFS, thus giving you enough time to confirm that concurrent READ's are happening.

If you do all this and then run your test program, you should get output similar to the following:

```
memfs-x64[TID=1554]: FFFFE78A06C2CE50: >>Read 00000228ECAE7C50, Address=00000228ECA30000, Offset=0:0, Length=4096, Key=0
memfs-x64[TID=182c]: FFFFE78A06C82E50: >>Read 00000228ECAE7C50, Address=00000228ECA40000, Offset=0:400000, Length=4096, Key=0
memfs-x64[TID=0310]: FFFFE78A06806E50: >>Read 00000228ECAE7C50, Address=00000228ECA50000, Offset=0:800000, Length=4096, Key=0
memfs-x64[TID=0210]: FFFFE78A06DA0E50: >>Read 00000228ECAE7C50, Address=00000228ECA60000, Offset=0:c00000, Length=4096, Key=0
memfs-x64[TID=1554]: FFFFE78A06C2CE50: <<Read IoStatus=0[4096]
memfs-x64[TID=1554]: FFFFE78A03740E50: >>Read 00000228ECAE7C50, Address=00000228ECA30000, Offset=0:1000000, Length=4096, Key=0
memfs-x64[TID=1554]: FFFFE78A03740E50: <<Read IoStatus=c0000011[0]
memfs-x64[TID=1554]: FFFFE78A05EF4E50: >>Read 00000228ECAE7C50, Address=00000228ECA30000, Offset=0:1400000, Length=4096, Key=0
memfs-x64[TID=1554]: FFFFE78A05EF4E50: <<Read IoStatus=c0000011[0]
memfs-x64[TID=1554]: FFFFE78A06664E50: >>Read 00000228ECAE7C50, Address=00000228ECA30000, Offset=0:1800000, Length=4096, Key=0
memfs-x64[TID=1554]: FFFFE78A06664E50: <<Read IoStatus=c0000011[0]
memfs-x64[TID=1554]: FFFFE78A069F8E50: >>Read 00000228ECAE7C50, Address=00000228ECA30000, Offset=0:1c00000, Length=4096, Key=0
memfs-x64[TID=1554]: FFFFE78A069F8E50: <<Read IoStatus=c0000011[0]
memfs-x64[TID=1554]: FFFFE78A06DAEE50: >>Read 00000228ECAE7C50, Address=00000228ECA30000, Offset=0:2000000, Length=4096, Key=0
memfs-x64[TID=1554]: FFFFE78A06DAEE50: <<Read IoStatus=c0000011[0]
memfs-x64[TID=1554]: FFFFE78A06802E50: >>Read 00000228ECAE7C50, Address=00000228ECA30000, Offset=0:2400000, Length=4096, Key=0
memfs-x64[TID=1554]: FFFFE78A06802E50: <<Read IoStatus=c0000011[0]
memfs-x64[TID=1554]: FFFFE78A078F0E50: >>Read 00000228ECAE7C50, Address=00000228ECA30000, Offset=0:2800000, Length=4096, Key=0
memfs-x64[TID=1554]: FFFFE78A078F0E50: <<Read IoStatus=c0000011[0]
memfs-x64[TID=1554]: FFFFE78A05958E50: >>Read 00000228ECAE7C50, Address=00000228ECA30000, Offset=0:2c00000, Length=4096, Key=0
memfs-x64[TID=1554]: FFFFE78A05958E50: <<Read IoStatus=c0000011[0]
memfs-x64[TID=1554]: FFFFE78A07810E50: >>Read 00000228ECAE7C50, Address=00000228ECA30000, Offset=0:3000000, Length=4096, Key=0
memfs-x64[TID=1554]: FFFFE78A07810E50: <<Read IoStatus=c0000011[0]
memfs-x64[TID=1554]: FFFFE78A05E5CE50: >>Read 00000228ECAE7C50, Address=00000228ECA30000, Offset=0:3400000, Length=4096, Key=0
memfs-x64[TID=1554]: FFFFE78A05E5CE50: <<Read IoStatus=c0000011[0]
memfs-x64[TID=1554]: FFFFE78A05FA4E50: >>Read 00000228ECAE7C50, Address=00000228ECA30000, Offset=0:3800000, Length=4096, Key=0
memfs-x64[TID=1554]: FFFFE78A05FA4E50: <<Read IoStatus=c0000011[0]
memfs-x64[TID=1554]: FFFFE78A05FB8E50: >>Read 00000228ECAE7C50, Address=00000228ECA30000, Offset=0:3c00000, Length=4096, Key=0
memfs-x64[TID=1554]: FFFFE78A05FB8E50: <<Read IoStatus=c0000011[0]
memfs-x64[TID=0210]: FFFFE78A06DA0E50: <<Read IoStatus=0[4096]
memfs-x64[TID=182c]: FFFFE78A06C82E50: <<Read IoStatus=0[4096]
memfs-x64[TID=0310]: FFFFE78A06806E50: <<Read IoStatus=0[4096]
```

Note the 4 concurrent READs at the top of the log (I ran MEMFS on a VM with 2 CPU's and thus got 4 file system dispatcher threads).

Finally I include your test program with the modification of using `FILE_FLAG_NO_BUFFERING`:

```C
#include <iostream>
#include <Windows.h>
#include <fileapi.h>


int main()
{
    static long readOffset = 1024*1024 * 4;
    static int requests = 16;

	// generate a test file the first run
	if (GetFileAttributesW(L"v:\\foo.raw") == INVALID_FILE_ATTRIBUTES) {
		HANDLE whandle = ::CreateFileW(L"v:\\foo.raw", GENERIC_WRITE,
			0, NULL,
			CREATE_NEW, FILE_ATTRIBUTE_NORMAL | FILE_FLAG_NO_BUFFERING, NULL);
		if (INVALID_HANDLE_VALUE == whandle)
		{
			fprintf(stderr,
				"error %d\n",
				GetLastError());
			exit(1);
		} // if ends

		void* src = malloc(16 * 1024 * 1024);
		for (long i = 0; i < (1024 / 16) + 1; i++) {
			::WriteFile(whandle, src, 16 * 1024 * 1024, NULL, NULL);
		}

		::CloseHandle(whandle);
	}

	// open the file and completion port
    HANDLE handle = ::CreateFileW(L"v:\\foo.raw", GENERIC_READ ,
        0, 0,
        OPEN_ALWAYS, FILE_ATTRIBUTE_NORMAL | FILE_FLAG_OVERLAPPED | FILE_FLAG_NO_BUFFERING, 0);
    if (INVALID_HANDLE_VALUE == handle)
    {
		exit(1);
    } // if ends

	ULONG key = 1;
	HANDLE completionPort = CreateIoCompletionPort(handle, NULL, (ULONG_PTR)&key, 2);
	if (completionPort == NULL) {
		exit(1);
	}

	// issue 16 overlapped IO's
    void* buffers[16];
	OVERLAPPED ovr[16];

	int pendingCompletions = 0;
    for (int i = 0; i < 16; i++) {
		void* buffer = malloc(4096);
        buffers[i] = buffer;

        ovr[i].Offset = readOffset * i;
		ovr[i].OffsetHigh = 0;
		ovr[i].hEvent = NULL;

		boolean res = ReadFile(handle, buffer, 4096, NULL, &ovr[i]);
		int lastError = GetLastError();
		if (res) {
			// immediate read
			// no need to wait on overlapped

		}
		else {
			if (lastError == ERROR_IO_PENDING)
				pendingCompletions += 1;
		}
    }

	// collect the completed IO's
	LPOVERLAPPED CompletedOverlapped;

	for (int i = 0; i < pendingCompletions;  i++) {
        if (buffers[i] != NULL) {
			DWORD_PTR key;
			DWORD bytes;
			boolean bOK = GetQueuedCompletionStatus(completionPort, &bytes, &key, &CompletedOverlapped, INFINITE);

			if (!bOK) {
				fprintf(stderr,
					" %d\n",
					GetLastError());
				exit(1);
			}

		}
    }
}
```

**Comment by billziss-gh** [comment]
The fix for this problem has been merged into the master branch. The fix will be included in the next release. I am closing this issue.
